File: dotnet/src/MyDotey.QuantileEstimator/Kll/KllQuantileEstimator.cs
using System;
using System.Collections.Generic;

/**
 * @author koqizhao
 *
 * Mar 30, 2018
 * 
 * see also: https://github.com/edoliberty/streaming-quantiles
 * 
 */
namespace MyDotey.Quantile.Kll
{
    public class KllQuantileEstimator<T> : IQuantileEstimator<T>
    {
        private KllQuantileEstimatorConfig<T> _config;

        private List<Compactor> _compactors;
        private int _h;
        private int _size;
        private int _maxSize;

        public KllQuantileEstimator(KllQuantileEstimatorConfig<T> config)
        {
            if (config == null)
                throw new ArgumentNullException("config is null");

            _config = config;

            _compactors = new List<Compactor>();

            Grow();
        }

        public virtual void Add(T value)
        {
            _compactors[0].Add(value);
            _size++;
            if (_size >= _maxSize)
                Compress();
        }

        public virtual Dictionary<Double, T> Get(List<Double> quantiles)
        {
            List<ItemAndQuantile> itemsAndQuantiles = Cdf();
            if (itemsAndQuantiles.Count == 0)
                return null;

            Dictionary<Double, T> results = new Dictionary<Double, T>();

            int i = 0, j = 0;
            while (i < quantiles.Count && j < itemsAndQuantiles.Count)
            {
                Double quantile = quantiles[i];
                ItemAndQuantile itemAndQuantile = itemsAndQuantiles[j];
                if (itemAndQuantile.quantile < quantile)
                {
                    j++;
                    continue;
                }

                results[quantile] = itemAndQuantile.item;
                i++;
            }

            if (i < quantiles.Count)
            {
                T result = itemsAndQuantiles[j - 1].item;
                for (; i < quantiles.Count; i++)
                {
                    Double quantile = quantiles[i];
                    results[quantile] = result;
                }
            }

            return results;
        }

        protected virtual void Grow()
        {
            _compactors.Add(new Compactor(this));
            _h = _compactors.Count;

            int maxSize = 0;
            for (int i = 0; i < _h; i++)
            {
                maxSize += Capacity(i);
            }
            _maxSize = maxSize;
        }

        protected virtual int Capacity(int height)
        {
            int depth = _h - 1 - height;
            return (int)Math.Ceiling(Math.Pow(_config.C, depth) * _config.K) + 1;
        }

        protected virtual void Compress()
        {
            for (int i = 0; i < _compactors.Count; i++)
            {
                if (_compactors[i].Count >= Capacity(i))
                {
                    if (i + 1 >= _h)
                        Grow();

                    _compactors[i + 1].AddRange(_compactors[i].compact());
                    int size = 0;
                    for (int j = 0; j < _compactors.Count; j++)
                        size += _compactors[j].Count;
                    _size = size;

                    break;
                }
            }
        }

        protected virtual List<ItemAndQuantile> Cdf()
        {
            List<ItemAndWeight> itemsAndWeights = new List<ItemAndWeight>();
            for (int i = 0; i < _compactors.Count; i++)
            {
                for (int j = 0; j < _compactors[i].Count; j++)
                {
                    itemsAndWeights.Add(new ItemAndWeight(_compactors[i][j], (int)Math.Pow(2, i)));
                }
            }

            int totWeight = 0;
            foreach (ItemAndWeight itemAndWeight in itemsAndWeights)
                totWeight += itemAndWeight.weight;

            itemsAndWeights.Sort();

            List<ItemAndQuantile> Cdf = new List<ItemAndQuantile>();
            int cumWeight = 0;
            foreach (ItemAndWeight itemAndWeight in itemsAndWeights)
            {
                cumWeight += itemAndWeight.weight;
                Cdf.Add(new ItemAndQuantile(itemAndWeight.item, (double)cumWeight / totWeight));
            }

            return Cdf;
        }

        protected virtual void Merge(KllQuantileEstimator<T> other)
        {
            while (_h < other._h)
                Grow();

            for (int i = 0; i < other._h; i++)
                _compactors[i].AddRange(other._compactors[i]);

            int size = 0;
            for (int i = 0; i < _compactors.Count; i++)
                size += _compactors[i].Count;
            _size = size;

            while (_size >= _maxSize)
                Compress();
        }

        protected virtual int Rank(T value)
        {
            int r = 0;
            for (int i = 0; i < _compactors.Count; i++)
            {
                for (int j = 0; j < _compactors[i].Count; j++)
                {
                    if (_config.Comparer.Compare(_compactors[i][j], value) <= 0)
                        r += (int)Math.Pow(2, i);
                }
            }

            return r;
        }

        protected virtual List<ItemAndWeight> Ranks()
        {
            List<ItemAndWeight> itemsAndWeights = new List<ItemAndWeight>();
            for (int i = 0; i < _compactors.Count; i++)
            {
                for (int j = 0; j < _compactors[i].Count; j++)
                {
                    itemsAndWeights.Add(new ItemAndWeight(_compactors[i][j], (int)Math.Pow(2, i)));
                }
            }

            itemsAndWeights.Sort();

            List<ItemAndWeight> ranksList = new List<ItemAndWeight>();
            int cumWeight = 0;
            foreach (ItemAndWeight itemAndWeight in itemsAndWeights)
            {
                cumWeight += itemAndWeight.weight;
                ranksList.Add(new ItemAndWeight(itemAndWeight.item, cumWeight));
            }

            return ranksList;
        }

        protected class Compactor : List<T>
        {
            private static readonly Random Random = new Random();

            private KllQuantileEstimator<T> _kllQuantileEstimator;

            public Compactor(KllQuantileEstimator<T> kllQuantileEstimator)
            {
                _kllQuantileEstimator = kllQuantileEstimator;
            }

            public virtual List<T> compact()
            {
                Sort(_kllQuantileEstimator._config.Comparer);

                List<T> result = new List<T>();
                if (Random.NextDouble() < 0.5)
                {
                    while (Count >= 2)
                    {
                        int index = Count - 1;
                        RemoveAt(index);
                        result.Add(this[index - 1]);
                        RemoveAt(index - 1);
                    }
                }
                else
                {
                    while (Count >= 2)
                    {
                        int index = Count - 1;
                        result.Add(this[index]);
                        RemoveAt(index);
                        RemoveAt(index - 1);
                    }
                }

                return result;
            }
        }

        protected class ItemAndWeight : IComparable<ItemAndWeight>
        {
            public T item;
            public int weight;

            private KllQuantileEstimator<T> _kllQuantileEstimator;

            public ItemAndWeight(KllQuantileEstimator<T> kllQuantileEstimator)
            {
                _kllQuantileEstimator = kllQuantileEstimator;
            }

            public ItemAndWeight(T item, int weight)
            {
                this.item = item;
                this.weight = weight;
            }

            public virtual int CompareTo(ItemAndWeight o)
            {
                int r = _kllQuantileEstimator._config.Comparer.Compare(item, o.item);
                if (r != 0)
                    return r;

                return weight > o.weight ? 1 : (weight == o.weight ? 0 : -1);
            }
        }

        protected class ItemAndQuantile
        {
            public T item;
            public double quantile;

            public ItemAndQuantile(T item, double quantile)
            {
                this.item = item;
                this.quantile = quantile;
            }

            public override String ToString()
            {
                return String.Format("{{ item: {0}, quantile: {1} }}", item, quantile);
            }
        }
    }
}

File: java/quantile-estimator/src/main/java/org/mydotey/quantile/kll/KllQuantileEstimator.java
package org.mydotey.quantile.kll;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import org.mydotey.quantile.QuantileEstimator;

/**
 * @author koqizhao
 *
 * Mar 30, 2018
 * 
 * see also: https://github.com/edoliberty/streaming-quantiles
 * 
 */
public class KllQuantileEstimator<T> implements QuantileEstimator<T> {

    private KllQuantileEstimatorConfig<T> _config;

    private List<Compactor> _compactors;
    private int _h;
    private int _size;
    private int _maxSize;

    public KllQuantileEstimator(KllQuantileEstimatorConfig<T> config) {
        Objects.requireNonNull(config, "config is null");
        _config = config;

        _compactors = new ArrayList<>();

        grow();
    }

    @Override
    public void add(T value) {
        _compactors.get(0).add(value);
        _size++;
        if (_size >= _maxSize)
            compress();
    }

    @Override
    public Map<Double, T> get(List<Double> quantiles) {
        List<ItemAndQuantile> itemsAndQuantiles = cdf();
        if (itemsAndQuantiles.isEmpty())
            return null;

        HashMap<Double, T> results = new HashMap<>();

        int i = 0, j = 0;
        while (i < quantiles.size() && j < itemsAndQuantiles.size()) {
            Double quantile = quantiles.get(i);
            ItemAndQuantile itemAndQuantile = itemsAndQuantiles.get(j);
            if (itemAndQuantile.quantile < quantile) {
                j++;
                continue;
            }

            results.put(quantile, itemAndQuantile.item);
            i++;
        }

        if (i < quantiles.size()) {
            T result = itemsAndQuantiles.get(j - 1).item;
            for (; i < quantiles.size(); i++) {
                Double quantile = quantiles.get(i);
                results.put(quantile, result);
            }
        }

        return results;
    }

    protected void grow() {
        _compactors.add(new Compactor());
        _h = _compactors.size();

        int maxSize = 0;
        for (int i = 0; i < _h; i++) {
            maxSize += capacity(i);
        }
        _maxSize = maxSize;
    }

    protected int capacity(int height) {
        int depth = _h - 1 - height;
        return (int) Math.ceil(Math.pow(_config.getC(), depth) * _config.getK()) + 1;
    }

    protected void compress() {
        for (int i = 0; i < _compactors.size(); i++) {
            if (_compactors.get(i).size() >= capacity(i)) {
                if (i + 1 >= _h)
                    grow();

                _compactors.get(i + 1).addAll(_compactors.get(i).compact());
                int size = 0;
                for (int j = 0; j < _compactors.size(); j++)
                    size += _compactors.get(j).size();
                _size = size;

                break;
            }
        }
    }

    protected List<ItemAndQuantile> cdf() {
        List<ItemAndWeight> itemsAndWeights = new ArrayList<>();
        for (int i = 0; i < _compactors.size(); i++) {
            for (int j = 0; j < _compactors.get(i).size(); j++) {
                itemsAndWeights.add(new ItemAndWeight(_compactors.get(i).get(j), (int) Math.pow(2, i)));
            }
        }

        int totWeight = 0;
        for (ItemAndWeight itemAndWeight : itemsAndWeights)
            totWeight += itemAndWeight.weight;

        Collections.sort(itemsAndWeights);

        List<ItemAndQuantile> cdf = new ArrayList<>();
        int cumWeight = 0;
        for (ItemAndWeight itemAndWeight : itemsAndWeights) {
            cumWeight += itemAndWeight.weight;
            cdf.add(new ItemAndQuantile(itemAndWeight.item, (double) cumWeight / totWeight));
        }

        return cdf;
    }

    protected void merge(KllQuantileEstimator<T> other) {
        while (_h < other._h)
            grow();

        for (int i = 0; i < other._h; i++)
            _compactors.get(i).addAll(other._compactors.get(i));

        int size = 0;
        for (int i = 0; i < _compactors.size(); i++)
            size += _compactors.get(i).size();
        _size = size;

        while (_size >= _maxSize)
            compress();
    }

    protected int rank(T value) {
        int r = 0;
        for (int i = 0; i < _compactors.size(); i++) {
            for (int j = 0; j < _compactors.get(i).size(); j++) {
                if (_config.getComparator().compare(_compactors.get(i).get(j), value) <= 0)
                    r += Math.pow(2, i);
            }
        }

        return r;
    }

    protected List<ItemAndWeight> ranks() {
        List<ItemAndWeight> itemsAndWeights = new ArrayList<>();
        for (int i = 0; i < _compactors.size(); i++) {
            for (int j = 0; j < _compactors.get(i).size(); j++) {
                itemsAndWeights.add(new ItemAndWeight(_compactors.get(i).get(j), (int) Math.pow(2, i)));
            }
        }

        Collections.sort(itemsAndWeights);

        List<ItemAndWeight> ranksList = new ArrayList<>();
        int cumWeight = 0;
        for (ItemAndWeight itemAndWeight : itemsAndWeights) {
            cumWeight += itemAndWeight.weight;
            ranksList.add(new ItemAndWeight(itemAndWeight.item, cumWeight));
        }

        return ranksList;
    }

    protected class Compactor extends ArrayList<T> {

        private static final long serialVersionUID = 1L;

        public List<T> compact() {
            sort(KllQuantileEstimator.this._config.getComparator());

            List<T> result = new ArrayList<>();
            if (Math.random() < 0.5) {
                while (size() >= 2) {
                    int index = size() - 1;
                    remove(index);
                    result.add(remove(index - 1));
                }
            } else {
                while (size() >= 2) {
                    int index = size() - 1;
                    result.add(remove(index));
                    remove(index - 1);
                }
            }

            return result;
        }
    }

    protected class ItemAndWeight implements Comparable<ItemAndWeight> {
        public T item;
        public int weight;

        public ItemAndWeight(T item, int weight) {
            this.item = item;
            this.weight = weight;
        }

        @Override
        public int compareTo(ItemAndWeight o) {
            if (o == null)
                return 1;

            int r = KllQuantileEstimator.this._config.getComparator().compare(item, o.item);
            if (r != 0)
                return r;

            return weight > o.weight ? 1 : (weight == o.weight ? 0 : -1);
        }
    }

    protected class ItemAndQuantile {

        public T item;
        public double quantile;

        public ItemAndQuantile(T item, double quantile) {
            this.item = item;
            this.quantile = quantile;
        }

        @Override
        public String toString() {
            return String.format("{ item: %s, quantile: %s }", item, quantile);
        }
    }

}
